extensions/amp-story-360/0.1/amp-story-360.js

```diff
@@ -202,15 +272,188 @@ export class AmpStory360 extends AMP.BaseElement {
     container.appendChild(this.canvas_);
     this.applyFillContent(container, /* replacedContent */ true);
 
-    return Services.storyStoreServiceForOrNull(this.win).then(
-      (storeService) => {
+    // Initialize all services before proceeding
+    return Promise.all([
+      Services.storyStoreServiceForOrNull(this.win).then((storeService) => {
+        this.storeService_ = storeService;
+
+        storeService.subscribe(StateProperty.PAGE_SIZE, () =>
+          this.resizeRenderer_()
+        );
+
         storeService.subscribe(
-          StateProperty.PAGE_SIZE,
-          this.resizeRenderer_.bind(this),
-          false /* callToInitialize */
+          StateProperty.GYROSCOPE_PERMISSION_STATE,
+          (permissionState) => this.onPermissionState_(permissionState)
         );
+      }),
+
+      Services.localizationServiceForOrNull(this.element).then(
+        (localizationService) => {
+          this.localizationService_ = localizationService;
+        }
+      ),
+    ]).then(() => {
+      attr('controls') === 'gyroscope' && this.checkGyroscopePermissions_();
+      return Promise.resolve;
+    });
+  }
+
+  /**
+   * @param {string} permissionState
+   * @private
+   */
+  onPermissionState_(permissionState) {
+    if (permissionState === 'granted') {
+      this.enableGyroscope_();
+    } else if (permissionState === 'denied') {
+      this.gyroscopeControls_ = false;
+      this.togglePermissionClass_(true);
+    }
+  }
+
+  /** @private */
+  checkGyroscopePermissions_() {
+    //  Browser doesn't support DeviceOrientationEvent API.
+    if (typeof DeviceOrientationEvent === 'undefined') {
+      return;
+    }
+
+    // If browser doesn't require permission, enable gyroscope.
+    if (typeof DeviceOrientationEvent.requestPermission === 'undefined') {
+      this.enableGyroscope_();
+    }
+
+    // If DeviceOrientationEvent and permissions, build permission button.
+    if (typeof DeviceOrientationEvent.requestPermission === 'function') {
+      this.checkPermissionOnLoad_();
+    }
+  }
+
+  /**
+   * Creates a device orientation listener and sets gyroscopeControls_ state.
+   * If listener is not called in 1000ms, remove listener and resume animation.
+   * This happens on desktop browsers that support deviceorientation but aren't in motion.
+   * @private
+   */
+  enableGyroscope_() {
+    this.gyroscopeControls_ = true;
+    this.togglePermissionClass_(true);
+
+    this.win.addEventListener('deviceorientation', (e) => {
+      if (this.isReady_ && this.isPlaying_) {
```

Comment:
> on long press we pause the story and set isPlaying_

I don't think that's built yet. It's mentioned in #29023 but doesn't appear to be implemented. 

> do we want to still be able to move the device to explore the media?

It could be cool to pause a video and look around the paused frame. I feel like we need to build video support so we can prototype that UX.

Reply:
We definitely want to implement it before the launch, it's tracked here: #29593
You wouldn't be able to look around when long pressing to pause once it's implemented. Should we add a new boolean to know whether the 360 is on the active page? Happy to help you figure out how to do so, I believe there are various acceptable ways to accomplish it.

Reply:
Sounds good. I set up a listener to check the active page ID against the parent page ID.